### PR TITLE
Spawn vscode on demand

### LINKED ISSUE #4481: Spawn VS Code on demand to avoid overhead
Introduced in #4414

@code-asher may have more details.

### PR DESCRIPTION
**Pull request by TeffenEllis**
Fixes #4481

### CODE REVIEW

**Review by code-asher — APPROVED** [review]
Thank you for doing this!  I know it sucks to maintain legacy behavior.  Once plugins are removed we can revert this behavior as well if we want. :smile:

File: src/node/routes/index.ts [hunk 1]
```diff
@@ -138,20 +138,12 @@ export const register = async (app: App, args: DefaultedArgs): Promise<Disposabl
...
-      logger.warn("VS Server router may still be compiling.")
-    } else {
-      throw error
-    }
+  const vsServerRouteHandler = new CodeServerRouteWrapper()
+
+  // Note that the root route is replaced in Coder Enterprise by the plugin API.
+  for (const routePrefix of ["/", "/vscode"]) {
+    app.router.use(routePrefix, vsServerRouteHandler.router)
+    app.wsRouter.use(routePrefix, vsServerRouteHandler.wsRouter)
```
Comment (jsjoeio) [inline src/node/routes/index.ts:146]: so much cleaner 👏

File: src/node/routes/vscode.ts [hunk 2]
```diff
@@ -1,65 +1,105 @@
...
     }
+
     next()
-  })
+  }
 
-  router.all("*", ensureAuthenticated, (req, res, next) => {
-    req.on("error", (error: any) => {
+  private $proxyRequest: express.Handler = async (req, res, next) => {
+    // We allow certain errors to propagate so that other routers may handle requests
```
Comment (jsjoeio) [inline src/node/routes/vscode.ts:37]: 👏

❓ why start the name with `$`? is that some sort of special pattern?

File: src/node/routes/vscode.ts [hunk 3]
```diff
@@ -1,65 +1,105 @@
...
+     * @file ../../../vendor/modules/code-oss-dev/src/vs/server/main.js
+     */
+    const createVSServer = await loadAMDModule<CodeServerLib.CreateServer>(
+      "vs/server/remoteExtensionHostAgent",
+      "createServer",
+    )
+
+    try {
+      this._codeServerMain = await createVSServer(null, {
+        connectionToken: "0000",
```
Comment (jsjoeio) [inline src/node/routes/vscode.ts:80]: ❓ What's the `connectionToken` used for again?

**Review by jsjoeio — APPROVED** [review]
Nice work!

**Comment by code-asher** [comment]
One observation (this is not from this PR since I remember seeing this before) but I get a 404 when VS Code is still loading rather than the "VS Code may still be compiling" error.

File: src/node/routes/vscode.ts (hunk 2 above)
Comment (TeffenEllis) [reply src/node/routes/vscode.ts:37]: It’s something of a convention I’ve seen help with readability, similarly to prefixing private methods with `_`. Prefixing a method like `$proxyRequest` helps indicate a unique purpose compared to a more lengthy `proxyRequestHandler`

File: src/node/routes/vscode.ts (hunk 3 above)
Comment (TeffenEllis) [reply src/node/routes/vscode.ts:80]: Connection token is used by VS Code as a basic form of client-side security. At the moment we’ve disabled its usage, but I think it’s worth looking at again as upstream matures the feature.

Comment (code-asher) [reply src/node/routes/vscode.ts:80]: Related: https://github.com/cdr/code-server/issues/4479

**Review by jsjoeio — APPROVED** [review]

**Pull request merged**